### ISSUE TITLE
Minimal toolchains picking up the wrong module when it exists in both dummy and GCCcore

### ISSUE THREAD

**Issue by damianam** [hpcugent/easybuild-framework#1787]
It seems like EB picks up the wrong module during building. For instance, when `binutils` or `ncurses` are both available in `dummy` and `GCCcore`, `--dry-run` and `--extended-dry-run` will work properly and show the correct dependency. However, during the build, the referenced paths are the ones of the `dummy` packages, not the ones in `GCCcore`. This seems like a silent bug in most cases.

Attached are a dry run, an extended dry run, and a debug log. Note in the debug log how at some point early in the process it load the correct module (the one in `GCCcore`), but later on the wrong one replaces it.

In the dry run you can see how `binutils` seems to not be installed in `dummy`. That is due to the same bug. We have removed the module in `dummy` to force EB to load the correct one.

The lack of `flex` in `dummy` is just because we have hide it there, so it is safe to ignore.

[easybuild-Python-3.5.1-20160530.183325.DnIed.log.txt](https://github.com/hpcugent/easybuild-framework/files/289893/easybuild-Python-3.5.1-20160530.183325.DnIed.log.txt)
[Python-3.5.1-GCC-5.3.0-2.26.dry_run.txt](https://github.com/hpcugent/easybuild-framework/files/289894/Python-3.5.1-GCC-5.3.0-2.26.dry_run.txt)
[Python-3.5.1-GCC-5.3.0-2.26.extended_dry_run.txt](https://github.com/hpcugent/easybuild-framework/files/289895/Python-3.5.1-GCC-5.3.0-2.26.extended_dry_run.txt)

**Comment by boegel** [comment]
@damianam would you mind copy-pasting the contents of the `Core/GCC/5.3.0-2.26` module?

`binutils` gets loaded via that module, and since `GCCcore` gets (probably) loaded first, the `load binutils` picks up the module from `Compiler/GCCcore/5.3.0`.

However, isn't that by design? You want to use the binutils that was built with `GCCcore/5.3.0` whenever you're loading `GCC/5.3.0-2.25`, no?

**Comment by damianam** [comment]
> However, isn't that by design? You want to use the binutils that was built with GCCcore/5.3.0 whenever you're loading GCC/5.3.0-2.25, no?

Exactly. The problem is that this is not the binutils that gets picked up, the binutils that gets picked up is the one in `dummy`.

The GCC module seems correct to me, and behaves correctly when loading it manually:

```
help([[The GNU Compiler Collection includes front ends for C, C++, Objective-C, Fortran, Java, and Ada,
 as well as libraries for these languages (libstdc++, libgcj,...). - Homepage: http://gcc.gnu.org/]])

whatis([[Description: The GNU Compiler Collection includes front ends for C, C++, Objective-C, Fortran, Java, and Ada,
 as well as libraries for these languages (libstdc++, libgcj,...). - Homepage: http://gcc.gnu.org/]])

local root = "/usr/local/software/jureca/Stages/2016a/software/GCC/5.3.0-2.26"

conflict("GCC")

load("GCCcore/.5.3.0")

load("binutils/.2.26")
prepend_path("MODULEPATH", "/usr/local/software/jureca/Stages/2016a/modules/all/Compiler/GCC/5.3.0-2.26")
prepend_path("MODULEPATH", "/usr/local/software/jureca/Stages/2016a/modules/all/Compiler/mpi/GCC/5.3.0-2.26")

setenv("EBROOTGCC", root)
setenv("EBVERSIONGCC", "5.3.0")
setenv("EBDEVELGCC", pathJoin(root, "easybuild/Core-GCC-5.3.0-2.26-easybuild-devel"))

family("compiler")
```

**Comment by boegel** [comment]
@damianam OK, I mixed things up, sorry. :-)

Can you point out in the debug log exactly where/when/how it's picking up `binutils` from `Core`?

**Comment by damianam** [comment]
Well, in that log `binutils` is not picked from `Core` because we have removed the module precisely to avoid it due to this bug. The log shows the bug with `ncurses`.

In line 1197 the correct module is found. In line 1526 the dependencies for the easyconfig are marked to be loaded (or actually loaded, I am not sure). However, line 1661 shows the wrong `$EBROOTNCURSES`.

I think the problem is related to `$MODULEPATH`. Line 1575 shows this `MODULEPATH`:

`os.environ["MODULEPATH"] = "/usr/local/software/jureca/Stages/2016a/modules/all:/usr/local/software/jureca/Devel:/usr/local/software/jureca/Stages/2016a/modules/all/Compiler/mpi/GCC/5.3.0-2.26:/usr/local/software/jureca/Stages/2016a/modules/all/Compiler/GCC/5.3.0-2.26:/usr/local/software/jureca/Stages/2016a/modules/all/Compiler/GCCcore/5.3.0:/usr/local/software/jureca/Stages/2016a/modules/all/Core";`

Line 1577 shows this `MODULEPATH`:

`os.environ["MODULEPATH"] = "/usr/local/software/jureca/Stages/2016a/modules/all/Core:/usr/local/software/jureca/Stages/2016a/modules/all:/usr/local/software/jureca/Devel:/usr/local/software/jureca/Stages/2016a/modules/all/Compiler/mpi/GCC/5.3.0-2.26:/usr/local/software/jureca/Stages/2016a/modules/all/Compiler/GCC/5.3.0-2.26:/usr/local/software/jureca/Stages/2016a/modules/all/Compiler/GCCcore/5.3.0";`

For some reason `Core` is moved from the end to the beginning, and that causes the modules in `Core` to have precedence over the ones in `GCCcore` (or any other).

Based on the log and the lines surrounding that point I'd expect the issue to be in `modules.py`, between lines 613 and 641 of the 2.8.0 release

**Comment by ocaisa** [comment]
I have a feeling the real problem is coming from https://github.com/hpcugent/easybuild-framework/blob/master/easybuild/tools/modules.py#L350

I think it's bumping the Core to first place, it should only bump it to first in front of other non-EB related paths

**Comment by ocaisa** [comment]
Scratch that, the problem is https://github.com/hpcugent/easybuild-framework/blob/master/easybuild/tools/modules.py#L365 where it deliberately pushes the EB path to the front without considering the hierarchy

**Comment by boegel** [comment]
@ocaisa well, ok, but that's only to make sure the modules are available under their *full* name, e.g. `Core/ncurses/5.9`...

So, I think your first suggestion, that `<modules_prefix>/Core` is prepended to `$MODULEPATH` unconditionally is closer to the truth...

**Comment by ocaisa** [comment]
If you look at line 1590 from the debug log you'll see it bumped all the initial module paths (including `/usr/local/software/jureca/Devel` which just houses our developer module) in front of the paths created by the hierarchy and that comes from the function in L365

**Comment by ocaisa** [comment]
Is there a good reason for the reordering of https://github.com/hpcugent/easybuild-framework/blob/master/easybuild/tools/modules.py#L386 I think perhaps we could reorder/sanitize the module path initially only and not touch it again after since any subsequent changes are part of the eb process and not accidental

**Comment by ocaisa** [comment]
Ah, you're right, the problem is the removal in L362, why is that necessary?

**Comment by ocaisa** [comment]
We could explicitly exclude any module paths that pattern match to a subdir of `install_path(typ='modules')` in the shuffle triggered at L386? That would leave `install_path(typ='modules')/Core` where it is in the module path but still ensure that the full name modules are picked up first

**Comment by boegel** [comment]
@damianam I think #1795 fixes this bug...

**Comment by boegel** [comment]
fixed in #1795
